### PR TITLE
depr: [DEPR-130] Remove EdxRestApiClient usage in ecommerce-worker

### PR DESCRIPTION
**Pull request by ruzniaievdm**
YT: [OeX_Depr-29](https://youtrack.raccoongang.com/issue/OeX_Depr-29)
JIRA: [DEPR-130](https://openedx.atlassian.net/browse/DEPR-130)

### CODE REVIEW

File: ecommerce_worker/email/v1/braze/tests/test_tasks.py [hunk 1]
```diff
@@ -95,6 +96,8 @@ class SendEmailsViaBrazeTests(TestCase):
         'site_code': SITE_CODE,
     }
 
+    OAUTH_ACCESS_TOKEN_URL = f"{get_configuration('BACKEND_SERVICE_EDX_OAUTH2_PROVIDER_URL')}/access_token/"
```
Comment (idegtiarov) [inline ecommerce_worker/email/v1/braze/tests/test_tasks.py]: I would prefer to use `urljoin` function from the `urllib.parse` to ensure we build a proper URL with and without trailing slash in the configuration variable.

Here and in the URL builders below.

File: ecommerce_worker/email/v1/braze/tests/test_tasks.py [hunk 2]
```diff
@@ -281,9 +284,10 @@ def test_message_sent(self, mock_update_assignment):
         self.assertEqual(mock_update_assignment.call_count, 1)
 
     @responses.activate
-    @patch('ecommerce_worker.email.v1.utils.get_ecommerce_client')
-    def test_update_assignment_exception(self, mock_get_ecommerce_client):
+    @patch('ecommerce_worker.email.v1.utils.get_access_token')
+    def test_update_assignment_exception(self, mock_get_access_token):
         """ Verify a message is logged after an unsuccessful API call to update the status. """
+        responses.add(responses.POST, self.OAUTH_ACCESS_TOKEN_URL, status=400, body='{}')
```
Comment (idegtiarov) [inline ecommerce_worker/email/v1/braze/tests/test_tasks.py]: Could you please give me a clue why do we need this new line?

File: ecommerce_worker/email/v1/tests/test_utils.py [hunk 3]
```diff
@@ -29,6 +32,16 @@ def mock_ecommerce_assignment_email_api(self, body, status=200):
             body=json.dumps(body), content_type='application/json',
         )
 
+    def mock_access_token_api(self, status=200):
+        """ Mock POST requests to retrieve an access token for this site's service user. """
```
Comment (idegtiarov) [inline ecommerce_worker/email/v1/tests/test_utils.py]: Please follow the [OeX python style guide](https://edx.readthedocs.io/projects/edx-developer-guide/en/latest/style_guides/python-guidelines.html#docstrings), the single line docstring is recommended to be formated as a multiple-docstrings.

File: ecommerce_worker/email/v1/tests/test_utils.py [hunk 4]
```diff
@@ -29,6 +32,16 @@ def mock_ecommerce_assignment_email_api(self, body, status=200):
...
+        body = json.dumps(dict(access_token=self.ACCESS_TOKEN))
+        responses.add(
+            responses.POST,
+            self.OAUTH_ACCESS_TOKEN_URL,
+            status=status,
+            body=body
```
Comment (idegtiarov) [inline ecommerce_worker/email/v1/tests/test_utils.py]: It seems that we could use json param instead of the body, don't we?

File: ecommerce_worker/email/v1/utils.py [hunk 5]
```diff
@@ -18,17 +21,26 @@ def update_assignment_email_status(offer_assignment_id, send_id, status, site_co
...
-        'send_id': send_id,
-        'status': status,
-    }
+
+    api_url = urljoin(get_configuration('ECOMMERCE_API_ROOT', site_code=site_code), 'assignment-email/status/')
     try:
-        api_response = api.status().post(post_data)
-    except RequestException:
+        access_token = get_access_token()
+        headers = {'Authorization': f'JWT {access_token}'}
```
Comment (idegtiarov) [inline ecommerce_worker/email/v1/utils.py]: Since access_token is not used in any other places in the function we could call get_access_token() inside the f-string.

File: ecommerce_worker/email/v1/utils.py [hunk 6]
```diff
@@ -18,17 +21,26 @@ def update_assignment_email_status(offer_assignment_id, send_id, status, site_co
...
+        response = requests.post(
+            api_url,
+            data={
+                'offer_assignment_id': offer_assignment_id,
+                'send_id': send_id,
+                'status': status,
+            },
+            headers=headers
+        )
+        data = response.json()
```
Comment (idegtiarov) [inline ecommerce_worker/email/v1/utils.py:37]: It would be good to handle the request exception if it is not handled anywhere else.

File: ecommerce_worker/fulfillment/v1/tasks.py [hunk 7]
```diff
@@ -37,25 +38,38 @@ def fulfill_order(self, order_number, site_code=None, email_opt_in=False):
...
+    except HTTPError as exc:
+        status_code = exc.response.status_code
         if status_code == 406:
             # The order is not fulfillable. Therefore, it must be complete.
             logger.info('Order [%s] has already been fulfilled. Ignoring.', order_number)
             raise Ignore() from exc
+
         # Unknown client error. Let's retry to resolve it.
         logger.warning(
             'Fulfillment of order [%s] failed because of HttpClientError. Retrying',
```
Comment (idegtiarov) [inline ecommerce_worker/fulfillment/v1/tasks.py]: The logging message is not fully correct since we do not work with the HttpClientError any more

File: ecommerce_worker/email/v1/utils.py [hunk 8]
```diff
@@ -18,17 +21,26 @@ def update_assignment_email_status(offer_assignment_id, send_id, status, site_co
...
+        status = data['status']
```
Comment (idegtiarov) [inline ecommerce_worker/email/v1/utils.py]: Working with the handling exception is more expensive than getting a value by a key, so I would prefer to stay with the initial approach, or do you see some additional pros to checking for KeyError?

File: ecommerce_worker/fulfillment/v1/tests/test_tasks.py [hunk 9]
```diff
@@ -85,14 +103,16 @@ def test_fulfillment_unknown_client_error_retry_success(self):
...
+        with self.assertRaises(HTTPError):
             fulfill_order.delay(self.ORDER_NUMBER).get()
 
     @responses.activate
     def test_fulfillment_timeout(self):
         """Verify that the task raises an exception when fulfillment times out."""
+        self.mock_access_token_api()
         responses.add(responses.PUT, self.API_URL, status=404, body=exceptions.Timeout())
 
         with self.assertRaises(exceptions.Timeout):
```
Comment (idegtiarov) [inline ecommerce_worker/fulfillment/v1/tests/test_tasks.py:131]: We do not have exceptions anymore, BTW how do we handle Timeout error?

File: ecommerce_worker/utils.py [hunk 10]
```diff
@@ -46,19 +48,36 @@ def get_configuration(variable, site_code=None):
...
 
 
-def get_ecommerce_client(url_postfix='', site_code=None):
+def get_access_token():
     """
-    Get client for fetching data from ecommerce API.
-    Arguments:
-        site_code (str): (Optional) The SITE_OVERRIDES key to inspect for site-specific values
-        url_postfix (str): (Optional) The URL postfix value to append to the ECOMMERCE_API_ROOT value.
+    Returns an access token for this site's service user.
```
Comment (idegtiarov) [inline ecommerce_worker/utils.py:54]: Please pay attention to the multiple-docstrings recommended format, check the OeX python style guide for more details.

File: ecommerce_worker/utils.py [hunk 11]
```diff
@@ -46,19 +48,36 @@ def get_configuration(variable, site_code=None):
...
     """
-    ecommerce_api_root = get_configuration('ECOMMERCE_API_ROOT', site_code=site_code)
-    signing_key = get_configuration('JWT_SECRET_KEY', site_code=site_code)
-    issuer = get_configuration('JWT_ISSUER', site_code=site_code)
-    service_username = get_configuration('ECOMMERCE_SERVICE_USERNAME', site_code=site_code)
-    return EdxRestApiClient(
-        ecommerce_api_root + url_postfix, signing_key=signing_key, issuer=issuer, username=service_username)
+
+    oauth_access_token_url = f"{get_configuration('BACKEND_SERVICE_EDX_OAUTH2_PROVIDER_URL')}/access_token/"
+    response = requests.post(
```
Comment (idegtiarov) [inline ecommerce_worker/utils.py]: Please cover requests with the try-except block

File: ecommerce_worker/utils.py [hunk 12]
```diff
@@ -46,19 +48,36 @@ def get_configuration(variable, site_code=None):
...
+            'User-Agent': 'ecommerce-worker',
+        },
+        timeout=(REQUEST_CONNECT_TIMEOUT, REQUEST_READ_TIMEOUT)
+    )
+
+    try:
+        data = response.json()
+        access_token = data['access_token']
+
+    except (KeyError, json.decoder.JSONDecodeError) as json_error:
```
Comment (idegtiarov) [inline ecommerce_worker/utils.py]: The error we need to check during the json decoding should be `requests.exceptions.JSONDecodeError` but I do not think we will need it if we will call `raise_for_status()` before decoding.

**Review by idegtiarov — CHANGES_REQUESTED** [review]
Please find my comments, question, and requests inline.

File: ecommerce_worker/configuration/base.py [hunk 13]
```diff
@@ -53,8 +53,9 @@
 # END ORDER FULFILLMENT
 
 # AUTHENTICATION
-JWT_SECRET_KEY = 'insecure-secret-key'
-JWT_ISSUER = 'ecommerce_worker'
+BACKEND_SERVICE_EDX_OAUTH2_PROVIDER_URL = ...
```
Comment (dyudyunov) [inline ecommerce_worker/configuration/base.py]: I think we could use `'http://localhost:18000/oauth2'` here too
Maybe it will be less confusing for debugging when the correct production-level value was not set up

File: ecommerce_worker/email/v1/tests/test_utils.py [hunk 14]
```diff
@@ -18,9 +19,11 @@
 class EmailUtilsTests(TestCase):
     """Tests for email v1 utility functions."""
 
+    OAUTH_ACCESS_TOKEN_URL = f"{get_configuration('BACKEND_SERVICE_EDX_OAUTH2_PROVIDER_URL')}/access_token/"
+    ACCESS_TOKEN = 'FAKE-access-token'
+
     def mock_ecommerce_assignment_email_api(self, body, status=200):
         """ Mock POST requests to the ecommerce assignment-email API endpoint. """
-        responses.reset()
```
Comment (dyudyunov) [inline ecommerce_worker/email/v1/tests/test_utils.py]: This could be crucial, because if you will add several responses on the same URL with the same method -- the first one added will be working. Why did you remove it?

File: ecommerce_worker/fulfillment/v1/tasks.py [hunk 15]
```diff
@@ -37,25 +38,38 @@ def fulfill_order(self, order_number, site_code=None, email_opt_in=False):
...
         logger.info('Requesting fulfillment of order [%s].', order_number)
-        api.orders(order_number).fulfill.put(email_opt_in=email_opt_in)
-    except exceptions.HttpClientError as exc:
-        status_code = exc.response.status_code  # pylint: disable=no-member
+        access_token = get_access_token()
+
+        headers = {'Authorization': f'JWT {access_token}'}
+        response = requests.put(
+            api_url,
+            params={
```
Comment (dyudyunov) [inline ecommerce_worker/fulfillment/v1/tasks.py:54]: Is it correct to use params here? Or it should be data?

File: ecommerce_worker/email/v1/braze/tests/test_tasks.py (hunk 1 above)
Comment (ruzniaievdm) [reply ecommerce_worker/email/v1/braze/tests/test_tasks.py]: done

File: ecommerce_worker/email/v1/tests/test_utils.py (hunk 3 above)
Comment (ruzniaievdm) [reply ecommerce_worker/email/v1/tests/test_utils.py]: done

File: ecommerce_worker/email/v1/utils.py (hunk 5 above)
Comment (ruzniaievdm) [reply ecommerce_worker/email/v1/utils.py]: done

File: ecommerce_worker/fulfillment/v1/tasks.py (hunk 15 above)
Comment (ruzniaievdm) [reply ecommerce_worker/fulfillment/v1/tasks.py:54]: If use the `data` parameters you will get failed test `test_email_opt_in_parameter_sent` 
so i use the `params`

File: ecommerce_worker/configuration/base.py (hunk 13 above)
Comment (ruzniaievdm) [reply ecommerce_worker/configuration/base.py]: done

File: ecommerce_worker/email/v1/tests/test_utils.py (hunk 4 above)
Comment (ruzniaievdm) [reply ecommerce_worker/email/v1/tests/test_utils.py]: yes we can
updated

File: ecommerce_worker/email/v1/braze/tests/test_tasks.py (hunk 2 above)
Comment (ruzniaievdm) [reply ecommerce_worker/email/v1/braze/tests/test_tasks.py]: you are right it's redundant here

File: ecommerce_worker/email/v1/tests/test_utils.py (hunk 14 above)
Comment (ruzniaievdm) [reply ecommerce_worker/email/v1/tests/test_utils.py]: transfer this method to another function before current

File: ecommerce_worker/utils.py (hunk 11 above)
Comment (ruzniaievdm) [reply ecommerce_worker/utils.py]: done

File: ecommerce_worker/utils.py (hunk 10 above)
Comment (ruzniaievdm) [reply ecommerce_worker/utils.py:54]: fixed

File: ecommerce_worker/fulfillment/v1/tests/test_tasks.py (hunk 9 above)
Comment (ruzniaievdm) [reply ecommerce_worker/fulfillment/v1/tests/test_tasks.py:131]: added handle Timeout

File: ecommerce_worker/email/v1/utils.py (hunk 6 above)
Comment (ruzniaievdm) [reply ecommerce_worker/email/v1/utils.py:37]: fixed

**Comment by ruzniaievdm** [comment]
@idegtiarov @dyudyunov @UvgenGen updated!

File: ecommerce_worker/email/v1/tests/test_utils.py (hunk 14 above)
Comment (dyudyunov) [reply ecommerce_worker/email/v1/tests/test_utils.py]: got it, thanks

**Review by dyudyunov — APPROVED** [review]
Reviewed addressed comments, all looks fine

File: ecommerce_worker/email/v1/braze/tests/test_tasks.py [hunk 16]
```diff
@@ -95,6 +97,8 @@ class SendEmailsViaBrazeTests(TestCase):
         'site_code': SITE_CODE,
     }
 
+    OAUTH_ACCESS_TOKEN_URL = urljoin(get_configuration('BACKEND_SERVICE_EDX_OAUTH2_PROVIDER_URL'), 'access_token/')
```
Comment (idegtiarov) [inline ecommerce_worker/email/v1/braze/tests/test_tasks.py]: Let's add a trailing slash to the base_url param in the urljoin function:
```
    OAUTH_ACCESS_TOKEN_URL = urljoin(get_configuration('BACKEND_SERVICE_EDX_OAUTH2_PROVIDER_URL')+'/', 'access_token/')
```

File: ecommerce_worker/email/v1/tests/test_utils.py [hunk 17]
```diff
@@ -16,7 +17,12 @@
 
 @ddt.ddt
 class EmailUtilsTests(TestCase):
-    """Tests for email v1 utility functions."""
+    """
+    Tests for email v1 utility functions.
+    """
+
+    OAUTH_ACCESS_TOKEN_URL = urljoin(get_configuration('BACKEND_SERVICE_EDX_OAUTH2_PROVIDER_URL'), 'access_token/')
```
Comment (idegtiarov) [inline ecommerce_worker/email/v1/tests/test_utils.py]: ditto

File: ecommerce_worker/email/v1/utils.py [hunk 18]
```diff
@@ -18,17 +20,25 @@ def update_assignment_email_status(offer_assignment_id, send_id, status, site_co
...
         True or False based on model update status from Ecommerce api
     """
-    api = get_ecommerce_client(url_postfix='assignment-email/', site_code=site_code)
-    post_data = {
-        'offer_assignment_id': offer_assignment_id,
-        'send_id': send_id,
-        'status': status,
-    }
+
+    api_url = urljoin(get_configuration('ECOMMERCE_API_ROOT', site_code=site_code), 'assignment-email/status/')
```
Comment (idegtiarov) [inline ecommerce_worker/email/v1/utils.py]: ditto

File: ecommerce_worker/email/v1/utils.py (hunk 6 above)
Comment (idegtiarov) [reply ecommerce_worker/email/v1/utils.py:37]: raise_for_status() raises HTTPError you are checking RequestException only, could you please help me to get how do both exceptions relate to each other?

File: ecommerce_worker/fulfillment/v1/tasks.py [hunk 19]
```diff
@@ -29,33 +31,46 @@ def _retry_order(self, exception, max_fulfillment_retries, order_number):
...
 
     Arguments:
         order_number (str): Order number indicating which order to fulfill.
 
     Returns:
         None
     """
+
+    api_url = urljoin(
+        get_configuration('ECOMMERCE_API_ROOT', site_code=site_code),
```
Comment (idegtiarov) [inline ecommerce_worker/fulfillment/v1/tasks.py]: trailing slash to the base_url param is recommended to be added to prevent creating an incorrect url.

File: ecommerce_worker/fulfillment/v1/tests/test_tasks.py [hunk 20]
```diff
@@ -16,62 +18,87 @@
...
 class OrderFulfillmentTaskTests(TestCase):
-    """Tests of the order fulfillment task."""
+    """
+    Tests of the order fulfillment task.
+    """
     ORDER_NUMBER = 'FAKE-123456'
-    API_URL = '{root}/orders/{number}/fulfill/'.format(
-        root=get_configuration('ECOMMERCE_API_ROOT').strip('/'),
-        number=ORDER_NUMBER
+    API_URL = urljoin(get_configuration('ECOMMERCE_API_ROOT'), f'orders/{ORDER_NUMBER}/fulfill/')
```
Comment (idegtiarov) [inline ecommerce_worker/fulfillment/v1/tests/test_tasks.py]: ditto

File: ecommerce_worker/fulfillment/v1/tests/test_tasks.py [hunk 21]
```diff
@@ -16,62 +18,87 @@
...
+    OAUTH_ACCESS_TOKEN_URL = '{root}/access_token/'.format(
```
Comment (idegtiarov) [inline ecommerce_worker/fulfillment/v1/tests/test_tasks.py]: urljoin would be preferable here.

File: ecommerce_worker/fulfillment/v1/tests/test_tasks.py [hunk 22]
```diff
@@ -16,62 +18,87 @@
...
+        root=get_configuration('BACKEND_SERVICE_EDX_OAUTH2_PROVIDER_URL').strip('/')
     )
+    ACCESS_TOKEN = 'FAKE-access-token'
+
+    def mock_access_token_api(self, status=200):
+        """
+        Mock POST requests to retrieve an access token for this site's service user.
+        """
+        body = json.dumps(dict(access_token=self.ACCESS_TOKEN))
```
Comment (idegtiarov) [inline ecommerce_worker/fulfillment/v1/tests/test_tasks.py]: json param instead of body is preferable here as well

File: ecommerce_worker/utils.py [hunk 23]
```diff
@@ -46,19 +49,32 @@ def get_configuration(variable, site_code=None):
...
-        EdxRestApiClient object
+        str: JWT access token
     """
-    ecommerce_api_root = get_configuration('ECOMMERCE_API_ROOT', site_code=site_code)
-    signing_key = get_configuration('JWT_SECRET_KEY', site_code=site_code)
-    issuer = get_configuration('JWT_ISSUER', site_code=site_code)
-    service_username = get_configuration('ECOMMERCE_SERVICE_USERNAME', site_code=site_code)
-    return EdxRestApiClient(
-        ecommerce_api_root + url_postfix, signing_key=signing_key, issuer=issuer, username=service_username)
+    oauth_access_token_url = urljoin(get_configuration('BACKEND_SERVICE_EDX_OAUTH2_PROVIDER_URL'), 'access_token/')
```
Comment (idegtiarov) [inline ecommerce_worker/utils.py]: please add a trailing slash to the base_url param here as well

File: ecommerce_worker/utils.py [hunk 24]
```diff
@@ -46,19 +49,32 @@ def get_configuration(variable, site_code=None):
...
+            },
+            headers={
+                'User-Agent': 'ecommerce-worker',
+            },
+            timeout=(REQUEST_CONNECT_TIMEOUT, REQUEST_READ_TIMEOUT)
+        )
+        response.raise_for_status()
+        data = response.json()
+        access_token = data['access_token']
+    except KeyError as exc:
```
Comment (idegtiarov) [inline ecommerce_worker/utils.py]: Please handle the HTTPError exception instead. As for the access_token, what should be the default return in case the response is successfully got but key 'access_token' is not in the `data`?

**Review by idegtiarov — CHANGES_REQUESTED** [review]
Some more comments to improve and discuss, please take a look.

File: ecommerce_worker/utils.py (hunk 24 above)
Comment (ruzniaievdm) [reply ecommerce_worker/utils.py]: decided leave exception handling

**Comment by ruzniaievdm** [comment]
@idegtiarov  added a trailing slash and some update part getting of `access_token`

File: ecommerce_worker/fulfillment/v1/tasks.py (hunk 7 above)
Comment (idegtiarov) [reply ecommerce_worker/fulfillment/v1/tasks.py]: @ruzniaievdm please exchange `HttpClientError` with `HTTPError`

**Review by idegtiarov — CHANGES_REQUESTED** [review]
One minor nit other LGTM

File: ecommerce_worker/fulfillment/v1/tasks.py (hunk 7 above)
Comment (ruzniaievdm) [reply ecommerce_worker/fulfillment/v1/tasks.py]: @idegtiarov  `HttpClientError` is a deprecation class from `slumber.exceptions`
in case of use `RequestException` from `requests.exceptions` do not pass tests

**Review by idegtiarov — APPROVED** [review]
Thanks LGTM now!

File: ecommerce_worker/email/v1/utils.py (hunk 6 above)
Comment (ruzniaievdm) [reply ecommerce_worker/email/v1/utils.py:37]: HTTPError inherits RequestException
```
class HTTPError(RequestException):
    """An HTTP error occurred."""
```

Comment (idegtiarov) [reply ecommerce_worker/email/v1/utils.py:37]: I know that why we do not use Exception which is also inherited by RequestException instead?

**Review by idegtiarov — CHANGES_REQUESTED** [review]

File: ecommerce_worker/email/v1/utils.py (hunk 6 above)
Comment (ruzniaievdm) [reply ecommerce_worker/email/v1/utils.py:37]: you're right, changed to HTTPError

**Review by idegtiarov — APPROVED** [review]
